server/stream.go

```diff
@@ -286,20 +286,21 @@ type stream struct {
 
 	// TODO(dlc) - Hide everything below behind two pointers.
 	// Clustered mode.
-	sa         *streamAssignment // What the meta controller uses to assign streams to peers.
-	node       RaftNode          // Our RAFT node for the stream's group.
-	catchup    atomic.Bool       // Used to signal we are in catchup mode.
-	catchups   map[string]uint64 // The number of messages that need to be caught per peer.
-	syncSub    *subscription     // Internal subscription for sync messages (on "$JSC.SYNC").
-	infoSub    *subscription     // Internal subscription for stream info requests.
-	clMu       sync.Mutex        // The mutex for clseq and clfs.
-	clseq      uint64            // The current last seq being proposed to the NRG layer.
-	clfs       uint64            // The count (offset) of the number of failed NRG sequences used to compute clseq.
-	inflight   map[uint64]uint64 // Inflight message sizes per clseq.
-	lqsent     time.Time         // The time at which the last lost quorum advisory was sent. Used to rate limit.
-	uch        chan struct{}     // The channel to signal updates to the monitor routine.
-	compressOK bool              // True if we can do message compression in RAFT and catchup logic
-	inMonitor  bool              // True if the monitor routine has been started.
+	sa               *streamAssignment // What the meta controller uses to assign streams to peers.
+	node             RaftNode          // Our RAFT node for the stream's group.
+	catchup          atomic.Bool       // Used to signal we are in catchup mode.
+	catchups         map[string]uint64 // The number of messages that need to be caught per peer.
+	syncSub          *subscription     // Internal subscription for sync messages (on "$JSC.SYNC").
+	infoSub          *subscription     // Internal subscription for stream info requests.
+	clMu             sync.Mutex        // The mutex for clseq, clfs, inflight and inflightSubjects.
+	clseq            uint64            // The current last seq being proposed to the NRG layer.
+	clfs             uint64            // The count (offset) of the number of failed NRG sequences used to compute clseq.
+	inflight         map[uint64]uint64 // Inflight message sizes per clseq.
+	inflightSubjects map[string]uint64 // Inflight number of messages per subject.
```

Comment:
This could get pretty big in terms of in memory state..

Reply:
I assumed that because it's only storing in flight messages, that the size would be 0 most of the time and go maybe up to a few thousands temporarily during message bursts, so still acceptable IMHO.

I did hesitate adding the support for new per subject to the PR because I couldn't think of a good use case off the top of my head (doesn't mean one doesn't exist), but it needed to be handled and I thought it would be better to actually implement it rather than try to prevent it (which is not that easy/clean to do because what you want to forbid specifically is discard new+new per subject+sourcing (and you can add/remove sources at any time)).